File: finder/java/src/main/java/com/testsigma/flutter/FlutterFinder.java
package com.testsigma.flutter;


import com.testsigma.flutter.finders.*;
import org.openqa.selenium.remote.FileDetector;
import org.openqa.selenium.remote.RemoteWebDriver;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Flutter finder implementation of different type of flutter element finder
 *
 * @author renju
 * @version 0.0.4
 */
public class FlutterFinder implements ByValueKey, ByType, ByToolTip, ByText, ByAncestor, ByDescendant, BySemanticsLabel, PageBack {
    RemoteWebDriver driver;
    FileDetector fileDetector;
    private static final String FINDER_TYPE = "finderType";

    public FlutterFinder(RemoteWebDriver driver) {
        this.driver = driver;
        this.fileDetector = null;
    }

    @Override
    public FlutterElement byValueKey(String key) {
        FlutterElement flutterElement = new FlutterElement(Map
                .of(FINDER_TYPE, "ByValueKey",
                        "keyValueType", "String",
                        "keyValueString", key));
        flutterElement.setParent(driver);
        flutterElement.setFileDetector(fileDetector);
        return flutterElement;
    }

    @Override
    public FlutterElement byValueKey(int key) {
        FlutterElement flutterElement = new FlutterElement(Map
                .of(FINDER_TYPE, "ByValueKey",
                        "keyValueType", "int",
                        "keyValueString", key));
        flutterElement.setParent(driver);
        flutterElement.setFileDetector(fileDetector);
        return flutterElement;
    }

    @Override
    public FlutterElement byType(String type) {
        return new FlutterElement(Map.of(
                FINDER_TYPE, "ByType",
                "type", type
        ));
    }

    @Override
    public FlutterElement byToolTip(String toolTipText) {
        return new FlutterElement(Map.of(
                FINDER_TYPE, "ByTooltipMessage",
                "text", toolTipText
        ));
    }

    @Override
    public FlutterElement byText(String input) {
        return new FlutterElement(Map.of(
                FINDER_TYPE, "ByText",
                "text", input
        ));
    }

    @Override
    public FlutterElement byAncestor(FlutterElement of, FlutterElement matching, boolean matchRoot, boolean firstMatchOnly) {
        Map<String, Object> matchIdentifier = new HashMap<>(Map.of(
                FINDER_TYPE, "Ancestor",
                "matchRoot", matchRoot,
                "firstMatchOnly", firstMatchOnly
        ));
        matchIdentifier.put("of", of.getRawMap());
        matchIdentifier.put("matching", matching.getRawMap());
        return new FlutterElement(matchIdentifier);
    }

    @Override
    public FlutterElement byDescendant(FlutterElement of, FlutterElement matching, boolean matchRoot, boolean firstMatchOnly) {
        Map<String, Object> matchIdentifier = new HashMap<>(Map.of(
                FINDER_TYPE, "Descendant",
                "matchRoot", matchRoot,
                "firstMatchOnly", firstMatchOnly
        ));
        matchIdentifier.put("of", of.getRawMap());
        matchIdentifier.put("matching", matching.getRawMap());
        return new FlutterElement(matchIdentifier);
    }

    @Override
    public FlutterElement bySemanticsLabel(String label) {
        return new FlutterElement(Map.of(
                FINDER_TYPE, "BySemanticsLabel",
                "isRegExp", false,
                "label", label
        ));
    }

    @Override
    public FlutterElement bySemanticsLabel(Pattern label) {
        return new FlutterElement(Map.of(
                FINDER_TYPE, "BySemanticsLabel",
                "isRegExp", true,
                "label", label.toString()
        ));
    }

    @Override
    public FlutterElement pageBack() {
        return new FlutterElement(Map.of(FINDER_TYPE, "PageBack"));
    }
}

File: finder/kotlin/src/main/kotlin/pro/truongsinh/appium_flutter/FlutterFinder.kt
package pro.truongsinh.appium_flutter

import java.util.regex.Pattern

import org.openqa.selenium.remote.RemoteWebDriver
import org.openqa.selenium.remote.FileDetector
import pro.truongsinh.appium_flutter.finder.FlutterElement
import pro.truongsinh.appium_flutter.finder.ancestor as _ancestor
import pro.truongsinh.appium_flutter.finder.bySemanticsLabel as _bySemanticsLabel
import pro.truongsinh.appium_flutter.finder.byTooltip as _byTooltip
import pro.truongsinh.appium_flutter.finder.byType as _byType
import pro.truongsinh.appium_flutter.finder.byValueKey as _byValueKey
import pro.truongsinh.appium_flutter.finder.descendant as _descendant
import pro.truongsinh.appium_flutter.finder.pageBack as _pageBack
import pro.truongsinh.appium_flutter.finder.text as _text


public class FlutterFinder(driver: RemoteWebDriver) {
  private val driver = driver
  private val fileDetector = FileDetector({ _ -> null })
  fun ancestor(of: FlutterElement, matching: FlutterElement, matchRoot: Boolean = false, firstMatchOnly: Boolean = false): FlutterElement {
    val f = _ancestor(of, matching, matchRoot, firstMatchOnly)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun bySemanticsLabel(label: String): FlutterElement {
    val f = _bySemanticsLabel(label)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun bySemanticsLabel(label: Pattern): FlutterElement {
    val f = _bySemanticsLabel(label)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun byTooltip(input: String): FlutterElement {
    val f = _byTooltip(input)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun byType(input: String): FlutterElement {
    val f = _byType(input)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun byValueKey(input: String): FlutterElement {
    val f = _byValueKey(input)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun byValueKey(input: Int): FlutterElement {
    val f = _byValueKey(input)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun descendant(of: FlutterElement, matching: FlutterElement, matchRoot: Boolean = false, firstMatchOnly: Boolean = false): FlutterElement {
    val f = _descendant(of, matching, matchRoot, firstMatchOnly)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun pageBack(): FlutterElement {
    val f = _pageBack()
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
  fun text(input: String): FlutterElement {
    val f = _text(input)
    f.setParent(driver)
    f.setFileDetector(fileDetector)
    return f
  }
}
